src/disposeOnUnmount.js

```diff
@@ -38,9 +34,16 @@ export function disposeOnUnmount(target, propertyKeyOrFunction) {
         )
     }
 
+    // decorator's target is the prototype, so it doesn't have any instance properties like props
+    const isDecorator = !target.hasOwnProperty("props")
```

Comment:
Very fragile check. What if in future react devs will decide to move props to prototype?

Reply:
I agree that it isn't the best check ever but I don't know what else we can do to discriminate usage. Also, `props` is pretty much the one property that will always be tied to the instance (`props` on the prototype would mean that all instance would share the same props, which completely defeats their purpose), so I think that, even in the future, we're fine making that assumption. Moreover, since classes are pretty much "legacy" now, I doubt anything will change for them in the future, and even if this for whatever reason change, we could still change this check for something else at that (very improbable) moment.

Anyway, now that v6 is out, this PR is pretty much obsolete for me (I had hoped that it could have been merged before) since the new restrictions on the method/decorator are breaking a good chunk of my uses for it. I'll still try to update it when I have the time though

Reply:
Merging this one now, since it is relevant again after 6.10 :)

Replaced this specificy check with `typeof propertyKeyOrFunction === "string"`, since only when used as decorator we get a string here.

Reply:
This was my first idea too (and the one I thought about when I wrote the issue), but nothing prevents you to use `disposeOnUnmount(this, "myProperty")`... though I agree this isn't the way it's supposed to be used.

Anyway thanks a lot for merging the PR!